dlib/image_processing/box_overlap_testing.h

```diff
@@ -17,11 +17,11 @@ namespace dlib
         const drectangle& b
     ) 
     {
-        const double inner = a.intersect(b).area();
-        if (inner == 0)
+        const auto inner = a.intersect(b).area();
+        if (inner <= std::numeric_limits<double>::epsilon())
```

Comment:
```suggestion
        if (inner == 0)
```

Reply:
See https://github.com/davisking/dlib/pull/2735#discussion_r1120203182

Reply:
So, you insist on it?
Sadly I did not find a good source stating out why not to do it.

Reply:
Maybe this can be considered sufficient:
https://www.learncpp.com/cpp-tutorial/relational-operators-and-floating-point-comparisons/
Have a look at "Comparison of calculated floating point values can be problematic" under "Floating point equality" it's explicitly said that == needs to be avoided if floating point values are calculated by any chance.

Reply:
It really should compare to 0.  The point of this is just to avoid division by 0.   Also don't use `auto`.  In general mostly don't use `auto`. It's a convenience for the writer of the code.  But code will be read many more times than it is written, so should be organized to make it as clear as possible.  Auto is good when the type itself is really irrelevant or harmful for readability.  Like when capturing a lambda or big complicated template type.

Reply:
Sorry, I know I tend to abuse auto...

Reply:
Well, my reaction in short:
I could live with dropping auto for double, allthough I do not fully aggree with your statement, Davis. But that's fine.
What I cannot accept is comparison for equality of a potentially (and it's quite likely in this context) calculated floating point variable with zero.
So, if both of you intend on this type of comparison I will sadly have to delete my Pull Request, as I'm not willing to be git blamable anywhere close to such a line of code.

Reply:
First of all, if you don't touch that line, you'll not be git blamable, so that should not be a problem.

Secondly, even if comparing two doubles for equality can be problematic, and even if you don't agree with comparing to 0 in particular is OK, in the `drectangle::area()` method you'll see that we are returning **exactly** 0 if the rectangle is empty.

From what I understand in that method code, there's no way to compute a 0 value because the area of the rectangle will always be either ≥ 1 (computed) or 0 (exact literal value).

Reply:
I understood the fact, that the area function will return exactly 0 if (and only if) either width or height are exactly zero. But as drectangles can be scaled by multiplication or division there are cases for which the widht or height can less or equal to double's epsilon, for which all the checks will fail and an area less 1 will be calculated.
So why activley ignoring one of programming's standard rules to **not** compare floating point variable with 0?

Reply:
Again, the width and height methods always add 1 to the difference between the edges, so even if you scale the rectangle down, you'll get 1 for the area. You can read more about this here: https://github.com/davisking/dlib/issues/463

Reply:
I give up.
The main thing is that the function really calculates the IoU.

Reply:
> Sorry, I know I tend to abuse auto...

No worries :)

Reply:
There really isn't anything wrong with comparing to 0.  People get into trouble when they write code like this though
```
void foo(double a, double b) 
{
    double val = a + b + b; // might end up being equal to just a if b is less than a*epsilon.
    double val2 = val - 2*b; // And say 2*b > a*epsilon, then this will make val2 != val.  So now val2 != a.
    if (val2 == a) 
    {
         // this might never happen depending on the value of b.  This foo function is just fundamentally numerically unsound.
    }
}
```

But that isn't the kind of situation here. And in general you definitely do not want to use the machine epsilon as a proxy for "it's zero".  For double it's only about 2e-16.  But values of double like 1e-30 are perfectly legit and meaningful.  The machine epsilon is really just a convenient way to ask "how much precision is in the mantissa?".  It's not telling us how much precision we have left in the exponent.

So for example, if you wanted to compute the ratio of two lengths `a/b` you want to do just `a/b`.  And if `b` can be 0 you have to decide what you want `a/0` to mean.  So 
```
if (b == 0) return whatever
else return a/b;
```
is what you do.  If instead it was this
```
if (b < std::numeric_limits<double>::epsilon()) return whatever
else return a/b;
```
then the wrong value would be obtained for values like 1e-20/1e-19.  Which should be 10.0, not `whatever`.

In the above foo example, you would use the epsilon in a setup like this
```
void foo(double a, double b) 
{
    double val = a + b + b;
    double val2 = val - 2*b;
    if (std::abs(val2 - a) <= std::numeric_limits<double>::epsilon() * 2*b) 
    {
         // This will execute.
    }
}
```